IPNI_HTTP_PROVIDER.md

```diff
@@ -0,0 +1,328 @@
+
+For more information on caching headers, you can refer to [RFC5861](https://datatracker.ietf.org/doc/html/rfc5861).
+
+#### Response
+
+The response from the `/ipni/v1/ad/head` endpoint includes the following fields:
+
+- **`head`** (required): The CID of the latest advertisement published by the provider.
+- **`topic`** (optional): The topic name on which the advertisement is published. If not specified, the default value
+  of `/indexer/ingest/mainnet` is assumed.
+- **`pubkey`** (required): The serialized public key of the provider in protobuf format using the libp2p standard. The
+  public key can be marshalled using
+  the [`crypto.MarshalPublicKey`](https://pkg.go.dev/github.com/libp2p/go-libp2p-core/crypto#MarshalPublicKey) function.
+- **`sig`** (required): The signature associated with the `head` CID, obtained by concatenating the bytes of the `head`
+  CID and the UTF-8 bytes of the `topic` (if present). The signature is verified against the `pubkey`.
+
+Please note that the response provides the necessary information to validate the authenticity of the advertisement CID
+and verify its integrity using the provider's public key and the associated signature.
+
+The following snippet represents the IPLD schema of the signed head advertisement:
+
+```ipldsch
+type SignedHead  struct {
+    head   Link
+    topic  optional String
+    pubkey Bytes
+    sig    Bytes
+}
+```
+
+#### Example
+
+The following represents an abbreviated example of head advertisement response in `DAG-JSON` encoding:
+
+```json
+{
+  "head": {
+    "/": "baguqeerazx6qhbdzckrnuwvl3reqnrpysz3ry5qtdxzpo23losoeiywij65a"
+  },
+  "topic": "/indexer/ingest/devnet",
+  "pubkey": {
+    "/": {
+      "bytes": "CAASpgIwggEiMA0G..."
+    }
+  },
+  "sig": {
+    "/": {
+      "bytes": "LrZiicKdqDqkG2UFR..."
+    }
+  }
+}
+```
+
+### GET `/ipni/v1/ad/{CID}`
+
+This endpoint retrieves the content associated with an advertisement CID and its entries. 
+The `CID` specified in the URL parameter must match the response body, as the data returned by this endpoint is immutable.
+
+To ensure proper caching and immutability, implementers must include the following response header:
+
+```
+Cache-Control: public, max-age=29030400, immutable
+```
+
+By setting this response header, it instructs client-side caches and intermediate proxies to store the response for a long duration (`max-age=29030400`), as well as treat it as immutable. This helps improve performance and ensures that the same content is served consistently for the specified `CID`.
+
+### Response
+
+The response from the `/ipni/v1/ad/{CID}` endpoint returns the data associated to the advertisement CID and its embedded links, including `Entries`. See [IPNI Specification/Advertisements](IPNI.md#advertisements) for the advertisement and entries schema. 
```

Comment:
It is not clear whether:
1. The entries are included in the response to a request for an advertisement CID
2. One separate request for all the entries is required (unless max content length exceeded)
3. A separate request is required for each block of entries.

Option 2 is the best choice IMO.  Not 1, because a removal advertisement later in the chain may mean that the indexer does not need to download the entries. Also, a utility inspecting the chain may not want the entries. Not 3, because the provider should be able to stream these to the indexer (nd-json encoded).

It should also be stated explicitly that the same endpoint is used to request advertisements as well as entries blocks.

Suggest:
> Entries must be retrieved using a separate request for the CID of the next entries block to fetch. The response contains one or more blocks of entries, up to the maximum content length. The first block in the response must match the CID requested (see entries verification). If more entries remain to be fetched, this is indicated by the last included entries block having a non-empty "Next" field. A subsequent request for the CID specified in the "Next" field retrieves the next set of entries blocks. This continues until the last block has an empty "Next" field.

Reply:
I think that blocks should be capable of being streamed, using nd-json encoding. This would remove the need to paginate or to create hold a very large response body in memory.

Reply:
* Encoding of body is a separate issue and HTTP offers enough extensibility with Conent-Type and Accept headers to address it. 
* Entries are retrieved by their CID just like an advertisement via `/ipni/v1/ad/{CID}`. The word `ad` in the URI does not imply that it only returns IPLD nodes of type `Advertisement`. Instead it means the endpoint is dedicated to serving ad related content. This is stated [here](https://github.com/ipni/specs/pull/18/files#diff-499b31ce9fbe231ed9ab3dfc01be2bb456792a9edeb6f1df749e9b6c126db871R220).
* I don't think we want to have a separate URI for each IPNI model because to get a CID that one does not know the type of one would have to try all endpoints. It would also defeat the purpose of CIDs a bit?
* This would also solve the pagination problem in the same way that it is currently solved by `Next` link in entries traversal.

Does that make sense? 
In the meantime, I will refine the test to state that the endpoint returns all ad related content if [this](https://github.com/ipni/specs/pull/18/files#diff-499b31ce9fbe231ed9ab3dfc01be2bb456792a9edeb6f1df749e9b6c126db871R220) does not suffice.

Reply:
> Encoding of body is a separate issue

OK, the specifics of what is put into the body can go into a separate (OpenAPI) spec. 

> Entries are retrieved by their CID just like an advertisement via /ipni/v1/ad/{CID}

Thanks. Just wanted to clarify that.

> I don't think we want to have a separate URI for each IPNI model because to get a CID that one does not know the type of one would have to try all endpoints. It would also defeat the purpose of CIDs a bit?

The point of having a separate URI is to explicitly identify what type of CID is being asked for. If using the ad/ endpoint then the provider only needs to look in the place where ads are kept, and if using an entries/ endpoint then only look where the entries data is kept.  I cannot think of a time when the requestor will not know whether they are requesting an ad or an entry block.  

Anyway, this is OK to use the same endpoint for both. It just means that the provider will not know what kind of object the CID is retrieving, which is find. The provider just has to retrieve it.

 > This would also solve the pagination problem in the same way that it is currently solved by Next link in entries traversal.
 
 The question is about whether there is, or should be, any expectation about 1 entry block per request or about n entry blocks per request.